src/containerapp/azext_containerapp/_dapr_utils.py

```diff
@@ -0,0 +1,338 @@
+        component["properties"]["metadata"] = metadata_items
+
+        return component
+
+    @staticmethod
+    def _is_component_created_by_daprcomponentinit(component_def) -> bool:
+        """
+        Check if the component was created by dapr-component init.
+
+        :param component_def: component definition to check
+
+        :return: True if the component was created by DaprUtils, False otherwise
+        """
+        from .custom import safe_get
+
+        if component_def is None:
+            raise ValidationError("Component definition cannot be None.")
+
+        service_binding_metadata_created_by = safe_get(
+            component_def,
+            "properties",
+            "serviceComponentBind",
+            "metadata",
+            DAPRCOMPONENTINIT_SERVICEBIND_METADATA_CREATEDBY_KEY,
+        )
+
+        # If the component was created by dapr-component init, it will have the metadata key.
+        # This can be created by the CLI or another source (like portal), so we skip the check for the value.
+        if service_binding_metadata_created_by is not None:
+            return True
+
+        return False
+
+    @staticmethod
+    def create_dapr_component_with_service_binding(
+        cmd,
+        component_name: str,
+        component_type: str,
+        service_type: str,
+        service_name: str,
+        service_id: str,
+        resource_group_name: str,
+        environment_name: str,
+        component_metadata: Dict[str, str] = None,
+    ):
+        """
+        Create a Dapr component with a service binding if it does not exist.
+
+        :param component_name: name of the Dapr component to create, e.g. statestore-redis
+        :param component_type: type of the Dapr component to create, e.g. state or pubsub
+        :param service_type: type of the service to bind to, e.g. redis or kafka
+        :param service_name: name of the service to bind to, e.g. dapr-redis
+        :param service_id: id of the service to bind to, e.g. /subscriptions/.../dapr-redis
+        :param component_metadata: metadata to add to the Dapr component, e.g. {"key": "value"}
+
+        :return: Dapr component definition of the component (whether it was created or not)
+        """
+        if (
+            component_type not in DaprUtils.supported_dapr_components
+            or service_type not in DaprUtils.supported_dapr_components[component_type]
+        ):
+            raise ValidationError(
+                f"Dapr component type {component_type} with service type {service_type} is not supported."
+            )
+
+        # Check if the component already exists.
+        logger.debug("Looking up Dapr component %s", component_name)
+        component_def = None
+        try:
+            component_def = DaprComponentPreviewClient.show(
+                cmd, resource_group_name, environment_name, component_name
+            )
+        except Exception:  # pylint: disable=broad-except
```

Comment:
Please use `handle_non_404_status_code_exception` to avoid ignoring other types of errors.